File: client/src/app/shared/services/booking.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { environment } from "../../../environments/environment";
import { Observable } from "rxjs";
import { Booking } from "../models/Booking";
import { Room } from "../models/Room";

@Injectable({ providedIn: 'root' })
export class BookingService {
  private api = `${environment.apiBaseUrl}/bookings`;

  constructor(private http: HttpClient) {}

  createBooking(userId: number, isFreelancer: boolean, booking: any): Observable<Booking> {
    return this.http.post<Booking>(`${this.api}/user/${userId}?isFreelancer=${isFreelancer}`, booking);
  }

  getBookingsForUser(userId: number): Observable<Booking[]> {
  return this.http.get<Booking[]>(`${this.api}/user/${userId}`);
}

getAvailableRooms(date: string, startTime: string, endTime: string): Observable<Room[]> {
  const params = new HttpParams()
    .set('date', date)
    .set('startTime', startTime)
    .set('endTime', endTime);

  return this.http.get<Room[]>(`${this.api}/available-rooms`, { params });
}
}

File: server/src/main/java/com/coworkingspace/server/services/BookingService.java
package com.coworkingspace.server.services;

import com.coworkingspace.server.DTOs.BookingDTO;
import com.coworkingspace.server.DTOs.RoomDTO;
import com.coworkingspace.server.models.Room;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

public interface BookingService {
    BookingDTO createBooking(Long userId, boolean isFreelancer, BookingDTO bookingDTO);
    boolean checkRoomAvailability(Long roomId, LocalDate date, LocalTime from, LocalTime to);
    double calculatePrice(Room room, LocalTime from, LocalTime to);
    List<BookingDTO> getBookingsForUser(Long userId);
    List<RoomDTO> getAvailableRooms(LocalDate date, LocalTime startTime, LocalTime endTime);


}
